beluga_amcl/src/ndt_amcl_node.cpp

```diff
@@ -0,0 +1,915 @@
+// Copyright 2022-2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <beluga/algorithm/estimation.hpp>
+#include <beluga/algorithm/spatial_hash.hpp>
+#include <beluga/containers/tuple_vector.hpp>
+#include <beluga/motion/differential_drive_model.hpp>
+#include <beluga/motion/stationary_model.hpp>
+#include <beluga/primitives.hpp>
+#include <beluga/random/multivariate_normal_distribution.hpp>
+#include <beluga/sensor/ndt_sensor_model.hpp>
+#include <beluga/views/particles.hpp>
+#include <beluga_amcl/ndt_amcl_node.hpp>
+
+#include <tf2/convert.h>
+#include <tf2/utils.h>
+#include <tf2_ros/create_timer_ros.h>
+
+#include <beluga_ros/amcl.hpp>
+#include <chrono>
+#include <execution>
+#include <limits>
+#include <memory>
+#include <sophus/se2.hpp>
+#include <stdexcept>
+#include <string>
+#include <string_view>
+#include <tuple>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+
+#include <geometry_msgs/msg/pose_stamped.hpp>
+#include <lifecycle_msgs/msg/state.hpp>
+#include <range/v3/algorithm/transform.hpp>
+#include <range/v3/range/conversion.hpp>
+#include <rclcpp_components/register_node_macro.hpp>
+
+#include <beluga_ros/tf2_sophus.hpp>
+#include <variant>
+
+namespace beluga_amcl {
+
+namespace {
+
+constexpr std::string_view kDifferentialModelName = "differential_drive";
+constexpr std::string_view kOmnidirectionalModelName = "omnidirectional_drive";
+constexpr std::string_view kStationaryModelName = "stationary";
+constexpr std::string_view kNav2DifferentialModelName = "nav2_amcl::DifferentialMotionModel";
+constexpr std::string_view kNav2OmnidirectionalModelName = "nav2_amcl::OmniMotionModel";
+
+}  // namespace
+
+NdtAmclNode::NdtAmclNode(const rclcpp::NodeOptions& options)
```

Comment:
@serraramiro1 meta: it would seem like there is some room to deduplicate vast amounts of code between this node and the regular AMCL one with little effort. Many methods are verbatim copies.

Reply:
I like the changes, but I will say we still have a lot of duplicate code though. I'm OK with deferring to a follow-up ticket if we tackle this right after higher priority stuff (instead of forgetting about it).

Reply:
Created https://github.com/Ekumen-OS/beluga/issues/355